libs/traits/src/ops.rs

```diff
@@ -25,273 +22,464 @@ pub trait Signum: PartialOrd + Zero {
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(extrinsic_overflow(), Err(ArithmeticError::Overflow.into()));
+		/// assert_eq!(extrinsic_underflow(), Err(ArithmeticError::Underflow.into()));
+		/// ```
+		fn ensure_from(other: T) -> Result<Self, ArithmeticError> {
+			Self::try_from(other).map_err(|_| error::equivalent(other))
+		}
+	}
 
-	#[test]
-	fn fixed_point_support() {
-		assert_eq!(
-			FixedU128::from(3).ensure_sub(&FixedU128::from(1)),
-			Ok(FixedU128::from(2))
-		);
-		assert_eq!(
-			FixedU128::from(0).ensure_sub(&FixedU128::from(1)),
-			Err(ArithmeticError::Underflow.into())
-		);
+	/// Similar to [`TryInto`] but returning an `ArithmeticError` error.
+	pub trait EnsureInto<T: Signum>: TryInto<T> + Signum {
+		/// Performs the conversion returning an `ArithmeticError` if fails.
+		///
+		/// Similar to [`TryInto::try_into()`] but returning an `ArithmeticError` error
+		///
+		/// ```
+		/// use cfg_traits::ops::ensure::EnsureInto;
+		/// use sp_runtime::{DispatchResult, ArithmeticError, DispatchError};
+		///
+		/// fn extrinsic_overflow() -> DispatchResult {
+		///     let byte: u8 = 256u16.ensure_into()?;
+		///     Ok(())
+		/// }
+		///
+		/// fn extrinsic_underflow() -> DispatchResult {
+		///     let byte: i8 = (-129i16).ensure_into()?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(extrinsic_overflow(), Err(ArithmeticError::Overflow.into()));
+		/// assert_eq!(extrinsic_underflow(), Err(ArithmeticError::Underflow.into()));
+		/// ```
+		fn ensure_into(self) -> Result<T, ArithmeticError> {
+			self.try_into().map_err(|_| error::equivalent(self))
+		}
 	}
 
-	#[test]
-	fn per_thing_support() {
-		assert_eq!(
-			Perbill::from_percent(3).ensure_sub(&Perbill::from_percent(1)),
-			Ok(Perbill::from_percent(2))
-		);
-		assert_eq!(
-			Perbill::from_percent(0).ensure_sub(&Perbill::from_percent(1)),
-			Err(ArithmeticError::Underflow.into())
-		);
+	impl<T: TryFrom<S> + Signum, S: Signum> EnsureFrom<S> for T {}
+	impl<T: TryInto<S> + Signum, S: Signum> EnsureInto<S> for T {}
+
+	mod error {
+		use super::{ArithmeticError, NumSign, Signum};
+
+		pub fn equivalent<R: Signum>(r: R) -> ArithmeticError {
+			match r.signum() {
+				NumSign::Negative => ArithmeticError::Underflow,
+				NumSign::Positive => ArithmeticError::Overflow,
+			}
+		}
+
+		pub fn inverse<R: Signum>(r: R) -> ArithmeticError {
+			match r.signum() {
+				NumSign::Negative => ArithmeticError::Overflow,
+				NumSign::Positive => ArithmeticError::Underflow,
+			}
+		}
+
+		pub fn multiplication<L: Signum, R: Signum>(l: L, r: R) -> ArithmeticError {
+			match l.signum() != r.signum() {
+				true => ArithmeticError::Underflow,
+				false => ArithmeticError::Overflow,
+			}
+		}
+
+		pub fn division<N: Signum, D: Signum>(n: N, d: D) -> ArithmeticError {
+			if d.is_zero() {
+				ArithmeticError::DivisionByZero
+			} else {
+				match n.signum() != d.signum() {
+					true => ArithmeticError::Underflow,
+					false => ArithmeticError::Overflow,
+				}
```

Comment:
You could choose to call `multiplication` here to display that `division` for non-zero `d` is the same as calling multiplication. Up to you :)